### PR TITLE
Fix handling of null transducer models (#218)

### LINKED ISSUE #218: `get_transducer_absolute_filepaths` cannot handle null
Since the type of both `Transducer.registration_surface_filename` and `Transducer.transducer_body_filename` are given as `Optional[str]`, these should both be allowed to be set to `None`, or equivalently `null` in the json files. However `Database.get_transducer_absolute_filepaths` will fail to handle this case because it is checking to see whether the attributes are present rather than checking to see whether the attributes are non-None:

https://github.com/OpenwaterHealth/OpenLIFU-python/blob/1dcff787b74a3b301edbab2f52f6d52835bd0198/src/openlifu/db/database.py#L709-L734

To see the bug, try `get_transducer_absolute_filepaths` on a Transducer that has either of those attributes set to `None`/`null` -- it raises an exception because it attempts to construct a `Path` using `None`.

Besides fixing this issue, it would also be nice to make `get_transducer_absolute_filepaths` just return an empty dict if there is no absolute filepath info rather than raising an exception -- it should be perfectly fine for there to be no absolute filepath info because a Transducer with `Transducer.registration_surface_filename` and `Transducer.transducer_body_filename` set to `None` is perfectly valid as things currently stand.

### PR DESCRIPTION
**Pull request by ebrahimebrahim**
Closes #218 

Added a test that fails before and passes after this fix

### CODE REVIEW

**Comment by ebrahimebrahim** [comment]
added @sadhana-r for visibility and but you don't have to review unless you want to
@arhowe00 please review, ty!

File: src/openlifu/db/database.py [hunk 1]
```diff
@@ -707,29 +707,33 @@ def load_photoscan(self, subject_id, session_id, photoscan_id, load_data = False
         return photoscan
 
     def get_transducer_absolute_filepaths(self, transducer_id:str) -> dict:
```
Comment (arhowe00) [inline src/openlifu/db/database.py:709]: Are you sure that this should not be returned? Something tells me we should pass along the information that nothing is associated, since this function is intended to get the absolute filepaths, you could return the key still, but with an empty string or Optional[str] with None.

File: tests/test_database.py [hunk 2]
```diff
@@ -649,3 +649,38 @@ def test_write_transducer(example_database:Database, example_transducer: Transdu
...
+        registration_surface_path.touch()
+    if include_body_model:
+        transducer_body_path = Path(tmp_path/"test_db_files/example_transducer_body.obj")
+        transducer_body_path.parent.mkdir(parents=True, exist_ok=True)
+        transducer_body_path.touch()
+
+    example_database.write_transducer(
+        transducer=transducer,
+        registration_surface_model_filepath = registration_surface_path,
+        transducer_body_model_filepath = transducer_body_path,
```
Comment (arhowe00) [inline tests/test_database.py:672]:
```diff
-        registration_surface_model_filepath = registration_surface_path,
-        transducer_body_model_filepath = transducer_body_path,
+        registration_surface_model_filepath = str(registration_surface_path),
+        transducer_body_model_filepath = str(transducer_body_path),
```

**Review by arhowe00 — CHANGES_REQUESTED** [review]
I tested this on main and this branch with the M2 transducer, and everything seems to work. A few discussion points.

File: src/openlifu/db/database.py (hunk 1 above)
Comment (ebrahimebrahim) [reply src/openlifu/db/database.py:709]: empty path is probably not a good idea but I do agree with Optional[str] being a slight improvement since the type then clearly communicates to the user that they have to worry about the special case, whereas a missing key is something to worry about but the worry is not communicated in the type system

File: tests/test_database.py [hunk 3]
```diff
@@ -649,3 +649,38 @@ def test_write_transducer(example_database:Database, example_transducer: Transdu
...
     with pytest.raises(ValueError, match="file associated with transducer"):
         example_database.write_transducer(example_transducer, on_conflict=OnConflictOpts.OVERWRITE)
+
+@pytest.mark.parametrize("include_registration_surface", [True, False])
+@pytest.mark.parametrize("include_body_model", [True, False])
+def test_get_transducer_absolute_filepaths(example_database:Database, tmp_path:Path, include_registration_surface:bool, include_body_model:bool):
+    transducer = Transducer(id="transducer_for_test_get_transducer_absolute_filepaths")
+
+    registration_surface_path = None
+    transducer_body_path = None
```
[the remark this answers is not in the document]
Comment (ebrahimebrahim) [reply tests/test_database.py:659]: actually these need to be None because one of the test cases is to pass None into `write_transducer`

Is there a better way to tell the type checker what's going on? Maybe we can let it go for a unit test, but if there is a good way I am also interested

File: src/openlifu/db/database.py [hunk 4]
```diff
@@ -20,7 +20,7 @@
 from .user import User
 
 OnConflictOpts = Enum('OnConflictOpts', ['ERROR', 'OVERWRITE', 'SKIP'])
-
+PathLike = Union[str, os.PathLike]
```
Comment (ebrahimebrahim) [inline src/openlifu/db/database.py:23]: This is an attempt to make the type checker happier without fussing as much with path/str conversion when _using_ functions. LMK what you think of this. I stole the approach from MONAI: https://github.com/Project-MONAI/MONAI/blob/a09c1f08461cec3d2131fde3939ef38c3c4ad5fc/monai/config/type_definitions.py#L82

File: tests/test_database.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (arhowe00) [reply tests/test_database.py:659]: One thing you could do is make them `Optional[str]` and derive paths from that. Gonna make that proposed change. I think it might make the intent of the tes t more clear than parametrizing with the bools

File: src/openlifu/db/database.py (hunk 4 above)
Comment (arhowe00) [reply src/openlifu/db/database.py:23]: Yes this is an interesting approach and it works fine!

**Comment by arhowe00** [comment]
My type checker doesn't give any complaints about the one above, and we don't need to use empty path

File: tests/test_database.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ebrahimebrahim) [reply tests/test_database.py]: It looks good, thanks!
